src/php/Analyzer/TupleSymbol/InvokeSymbol.php

```diff
@@ -18,103 +18,109 @@ final class InvokeSymbol
+            $env,
             $f,
-            $arguments,
-            $x->getStartLocation()
+            $this->arguments($tuple, $env),
+            $tuple->getStartLocation()
         );
     }
 
+    private function globalMacro(Tuple $tuple, NodeEnvironment $env): Node
+    {
+        $this->analyzer->getGlobalEnvironment()->setAllowPrivateAccess(true);
+        $result = $this->analyzer->analyze($this->macroExpand($tuple, $env), $env);
+        $this->analyzer->getGlobalEnvironment()->setAllowPrivateAccess(false);
+
+        return $result;
+    }
+
     /** @return AbstractType|scalar|null */
-    private function macroExpand(Tuple $x, NodeEnvironment $env)
+    private function macroExpand(Tuple $tuple, NodeEnvironment $env)
     {
-        $tupleCount = count($x);
+        $tupleCount = count($tuple);
         /** @psalm-suppress PossiblyNullArgument */
-        $node = $this->analyzer->getGlobalEnvironment()->resolve($x[0], $env);
+        $node = $this->analyzer->getGlobalEnvironment()->resolve($tuple[0], $env);
         if ($node && $node instanceof GlobalVarNode) {
-            $fn = $GLOBALS['__phel'][$node->getNamespace()][$node->getName()->getName()];
+            $nodeName = $node->getName()->getName();
+            $fn = $GLOBALS['__phel'][$node->getNamespace()][$nodeName];
 
             $arguments = [];
             for ($i = 1; $i < $tupleCount; $i++) {
-                $arguments[] = $x[$i];
+                $arguments[] = $tuple[$i];
             }
 
             try {
                 $result = $fn(...$arguments);
-                $this->enrichLocation($result, $x);
+                $this->enrichLocation($result, $tuple);
 
                 return $result;
             } catch (Exception $e) {
-                throw new AnalyzerException(
-                    'Error in expanding macro "' . $node->getNamespace() . '\\' . $node->getName()->getName() . '": ' . $e->getMessage(),
-                    $x->getStartLocation(),
-                    $x->getEndLocation(),
+                throw AnalyzerException::withLocation(
+                    'Error in expanding macro "' . $node->getNamespace() . '\\' . $nodeName . '": ' . $e->getMessage(),
+                    $tuple,
                     $e
                 );
             }
         }
 
         if (is_null($node)) {
-            throw new AnalyzerException(
-                'Can not resolve macro',
-                $x->getStartLocation(),
-                $x->getEndLocation()
-            );
+            throw AnalyzerException::withLocation('Can not resolve macro', $tuple);
         }
 
-        throw new AnalyzerException(
-            'This is not macro expandable: ' . get_class($node),
-            $x->getStartLocation(),
-            $x->getEndLocation()
-        );
+        throw AnalyzerException::withLocation('This is not macro expandable: ' . get_class($node), $tuple);
     }
 
-    /** @param mixed $x */
+    /** @param Tuple|AbstractType $x */
```

Comment:
Tuple extends AbstractType. => $x can be AbstractType only

Reply:
Sorry, I was wrong. `$x` can be any Phel Type, because we walk recursively through the data. It should be `AbstractType|scalar|null`

Reply:
Done 🚵